package.json

```diff
@@ -20,13 +20,18 @@
     "lint:js:fix": "eslint . --fix",
     "lint": "npm-run-all --aggregate-output --continue-on-error --parallel 'lint:!(fix)'",
     "lint:fix": "npm-run-all --aggregate-output --continue-on-error --parallel lint:*:fix",
-    "test:ts": "tsc --noEmit",
     "start": "ember serve",
     "test": "ember test",
+    "test:ts": "tsc --noEmit",
```

Comment:
I know this is a sorting fix, but I'm curious to know your thoughts on naming this `lint:ts` instead of `test:ts`?

Specifically, this sort of type checking _feels_ like a lint thing and fits with other lint related tooling. I'd also prefer to not even run tests if `yarn lint` itself fails.

Reply:
There's an interesting semantic argument here that is probably worth hashing out as a community—maybe on [the RFC](https://github.com/typed-ember/ember-cli-typescript/pull/1158). Whether we encourage people to think of TS checks as *lints* or as *tests* gets at a pretty deep difference in how people approach it. Is a compiler error comparable to an ESLint error—i.e. something that could just be a matter of opinion? Or does it represent a *bug* the way a failing test does?

Reply:
Some theory arguments first (though these aren't my main focus/point):

* Linting errors represent bugs just as clearly as failing tests (obviously there are non-bug related lints e.g. prettier, but overall the things in most sets of `recommended` / `plugin:node/recommended` / etc configs **do** represent bugs). The primary differences are: linting is fast whereas tests are slow.
* This is essentially fallout of the ember-cli-typescript teams decision to use Babel as the primary mechanism for transpiling `.ts` files (note: I'm not saying this was wrong or bad in anyway). Babel literally **does not care** about the correctness of the TypeScript files, it merely removes the type annotations and happily compiles for runtime completely invalid TypeScript code. This effectively makes this category of errors no different than the category of errors that eslint's recommended checks against.

Now for the practical (and to me the primary) point:

Doing this as part of `yarn lint` means that TS failures **prevent** a test run. This might seem contrary to the line of thinking you had above (where you were logically arguing that TS failures are more important so they are basically test failures), but in practical terms `yarn test` (here and in all addons by default) runs linting **first**. The goal here is to identify _fundamental_ problems before wasting developer or CI time to run tests.

Reply:
Yep, and to be clear I'm not necessarily making an argument about which we should pick here, and if the best way to get to fast-failure is just to use the existing lint infrastructure, that may be absolutely the right thing to do—but I *do* have concerns about the *language* of "linting" and what it implies. I'm speaking there from experience over the last couple years helping folks with TS things: while I very much agree with you about what lint errors represent, a *lot* of people don't, and when those same people treat *type errors* as *lints*, it encourages them to just hand wave them if they don't like/disagree with/think they know better than the error.

I also think you're correct philosophically about the relationship between Babel + TS here! Moreover, the concerns I have about this in the big picture are not a concern for *this* repo, because there's shared philosophy here: lint errors *are errors*, whatever their source.

I do wonder whether we might want to use a name like `check` instead of `lint` more broadly. 🤔 

I'm going to add this as an unresolved question on the RFC, and would mildly prefer we discuss there with *this* as a motivating discussion source for it.

In the meantime, I'm happy enough to switch it to `lint` here for now in the interest of landing this; we can always tweak it later!

Reply:
Related: we have `test:types` currently, and that's run during the CI run as part of the *test* suite. I think we want that to *remain* a "test" thing that's part of CI only, *not* treat it as a lint step, because it's *not* a fast-fail operation. Thoughts?

Reply:
> I'm speaking there from experience over the last couple years helping folks with TS things: while I very much agree with you about what lint errors represent, a lot of people don't, and when those same people treat type errors as lints, it encourages them to just hand wave them if they don't like/disagree with/think they know better than the error.

I have definitely seen this as well, and I do agree that people are *way* to flippant about linting errors (but IMHO this is because they are using stupid/useless linting rules that just cause pain without justification 😡 ). 

> I'm going to add this as an unresolved question on the RFC, and would mildly prefer we discuss there with this as a motivating discussion source for it.

:+1: - Will move convo over to the RFC thread.

> In the meantime, I'm happy enough to switch it to lint here for now in the interest of landing this; we can always tweak it later!

👍

Reply:
- [Added this as an unresolved question on the RFC](https://github.com/typed-ember/ember-cli-typescript/pull/1158/commits/ec8f777c9b055168fb52466d0a8c60204b513a38), and am about to push the change here (along with the `typesVersions` fix).
- Totally agreed re: stupid/useless linting rules and their impact on people's habits. 😭

Reply:
> Related: we have test:types currently, and that's run during the CI run as part of the test suite. I think we want that to remain a "test" thing that's part of CI only, not treat it as a lint step, because it's not a fast-fail operation. Thoughts?

Yes, totally agree. It is running _much_ more than a "normal" `tsc --noEmit`, as it is testing each supported `tsc` version.